### PR TITLE
DAOS-7734 test: Wrap MPI code in unified interface

### PR DESCRIPTION
**Pull request by jolivier23**
Link daos test code with stub library that also implements serial version.  Using dlopen, the stub library
attempts to open libdpar_mpi.so.  If it can't find the library -or- can't load any of the symbols, it will print
a message indicating no MPI is present and run serially.

The examples were not modified.

Fix some compiler warnings

This allows us to ship optional MPI library but also ship all of the utilities without one.   Perhaps we also
should ship the source code for the MPI library so a user could build it on demand as needed for their own
MPI.

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

**Comment by jolivier23** [comment]
This is a proof of concept removing the direct MPI dependence

File: debian/daos-server-tests.install [hunk 1]
```diff
@@ -8,3 +8,7 @@ usr/bin/vea_ut
 usr/bin/srv_checksum_tests
 usr/bin/vos_tests
 usr/bin/vea_stress
+usr/bin/vos_perf
+usr/bin/daos_run_io_conf
+usr/bin/obj_ctl
+usr/bin/daos_gen_io_conf
```
Comment (brianjmurrell) [inline debian/daos-server-tests.install:14]: Do these need `libdpar.so` also?  I don't see how they are going to get it if so.  I don't see any dependency on `daos-client-tests` nor am I convinced there should be one.

File: src/tests/simple_common.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
+/**
+ * (C) Copyright 2016-2021 Intel Corporation.
```
Comment (brianjmurrell) [inline src/tests/simple_common.h:2]: Is this a new file?  Should the copyright only be 2021?

File: utils/rpms/daos.spec [hunk 3]
```diff
@@ -470,15 +462,17 @@ getent passwd daos_agent >/dev/null || useradd -s /sbin/nologin -r -g daos_agent
...
 %{_bindir}/daos_racer
 %{_bindir}/daos_test
 %{_bindir}/dfs_test
 %{_bindir}/jobtest
 %{_libdir}/libdts.so
+%{_libdir}/libdpar.so
+
+%files client-tests-openmpi
+%{_bindir}/crt_launch
+%{_libdir}/libdpar_mpi.so
```
Comment (brianjmurrell) [inline utils/rpms/daos.spec:475]: Same question about whether this is needed for the server tests.

**Review by brianjmurrell — COMMENTED** [review]
I probably don't know enough about MPI to review the bulk of this PR.  I do have a couple of questions about the packaging though.

I also wonder if any tests need updating/adding/removing (as obsolete perhaps) in [ci/rpm/test_daos_node.sh](https://github.com/daos-stack/daos/blob/7510701438ae3237a839d5733ecc472cd8b9c7f4/ci/rpm/test_daos_node.sh).

File: utils/rpms/daos.spec (hunk 3 above)
Comment (jolivier23) [reply utils/rpms/daos.spec:475]: It is needed for tests that are run with more than one rank.  Any tests that does mpirun ./test is fine but if they pass a hostlist of -np > 1, it is needed to also preload this library.  I don't think we run any server tests this way in any of our testing.

Do we use RPMs in our testing by default or should I run something explicitly to test that?

File: src/tests/simple_common.h (hunk 2 above)
Comment (jolivier23) [reply src/tests/simple_common.h:2]: yes, will fix

File: debian/daos-server-tests.install (hunk 1 above)
Comment (jolivier23) [reply debian/daos-server-tests.install:14]: They would need libdpar.so (or libdpar_mpi.so) to run, yes.   Maybe there is a more appropriate place to put that library.

I'm not sure I got the packaging right here which is why I added you to the patch.

Comment (brianjmurrell) [reply debian/daos-server-tests.install:14]: > They would need libdpar.so (or libdpar_mpi.so) to run, yes

Does that mean every consumer needs to specify an `LD_LIBRARY_PRELOAD` whether they want `libdpar.so` or `libdpar_mpi.so` or is the former a default without an `LD_LIBRARY_PRELOAD`?

> Maybe there is a more appropriate place to put that library.

There should be a "common" (i.e. `daos-tests-common`, probably) file that both packages depend on that would have the common _things_ (these libraries at least) in them

File: utils/rpms/daos.spec (hunk 3 above)
Comment (brianjmurrell) [reply utils/rpms/daos.spec:475]: > It is needed for tests that are run with more than one rank. Any tests that does mpirun ./test is fine but if they pass a hostlist of -np > 1, it is needed to also preload this library.

So basically `libdpar.so` is to allow tests to be written with MPI, but can be used when they don't actually run with multiple nodes?  Do we do a lot of this?

> Do we use RPMs in our testing by default

In functional testing, yes.

> or should I run something explicitly to test that?

I must have lost track.  Test what exactly?

File: debian/daos-server-tests.install (hunk 1 above)
Comment (jolivier23) [reply debian/daos-server-tests.install:14]: By default, everything is linked directly with libdpar.so.  If they want to use the MPI version, they do

mpirun -x LD_PRELOAD=libdpar_mpi.so [args] ./daos_test -i

The mpi version is just a drop-in replacement for the same interfaces.

File: utils/rpms/daos.spec (hunk 3 above)
Comment (jolivier23) [reply utils/rpms/daos.spec:475]: > 

Yes, I would say most of the runs we do are on a single node, single rank and do not need MPI.  We only have a handful of tests that need libdpar_mpi.so to be run in CI.

File: debian/daos-server-tests.install (hunk 1 above)
Comment (jolivier23) [reply debian/daos-server-tests.install:14]: @brianjmurrell I may need your help with the spec changes if we do more there

File: src/common/dpar_stub.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	ACTION(barrier, arg)		\
+	ACTION(rank, arg)		\
+	ACTION(size, arg)		\
+	ACTION(reduce, arg)		\
+	ACTION(gather, arg)		\
+	ACTION(allreduce, arg)		\
+	ACTION(allgather, arg)		\
+	ACTION(bcast, arg)
+
+#define LOAD_SYM(name, fail)							\
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/common/dpar_stub.c:41]: They should when they are part of a foreach

**Comment by jolivier23** [comment]
@johannlombardi what do you think I should do with this?

The nice thing is it would remove direct dependence on MPI for some of our utilities so you can run this without any MPI.  The downside is it does change packaging some and it would cause a conflict with one of my other PRs.  But I've refreshed it a few times and keeping hitting spec file conflicts

**Comment by jolivier23** [comment]
Closing this in favor of a new patch with clean history
https://github.com/daos-stack/daos/pull/8127

**Pull request closed without merging**